File: patterns/01_factory-method/ex2_imageDecoder.go
package main

import "fmt"

// Step 1: define a get method for all decoders
type DecodeMethod interface {
	getDecodeImage(image string) DecodeImage
}

// Step 2: Define decode type to get a suitable method
type DecodeType int

// Step 2.1: Define table to summary all decoders
const (
	GIF DecodeType = iota
	JPEG
)

// Step 3: Create a concrete class
type decodeImage struct {
	image string
}

// Step 3.1: Implement Interfaces to build from concrete class
type DecodeImage interface {
	init(image string)
	toString() string
}

// Step 3.2: Implement methods from interface
func (d *decodeImage) init(image string) {
	d.image = image
}

func (d decodeImage) toString() string {
	return d.image + ": is decoded"
}

// Step 4: Define struct decoder method
type GifReader struct {
	// Fix issue 1: using next to structure
	next decodeImage
}
type JpegReader struct {
	// Fix issue 1: using next to structure
	next decodeImage
}

// Step 4: Implement
func (g *GifReader) getDecodeImage(image string) DecodeImage {
	fmt.Println("get gif decoder success with image " + image)
	return &g.next
}

func (j *JpegReader) getDecodeImage(image string) DecodeImage {
	fmt.Println("get jpeg decoder success with image" + image)

	return &j.next
}

// Step 5: Select decoder from table
// Issue 1: How to fix, missing a return type from struct
func GetDecodeMethod(t DecodeType) DecodeMethod {
	switch t {
	case GIF:
		return new(GifReader)
	case JPEG:
		return new(JpegReader)
	default:
		return nil
	}
}

func main() {

	image := "test.gif"
	format := "gif"

	if format == "gif" {
		reader := GetDecodeMethod(GIF)
		fmt.Print(reader.getDecodeImage(image).toString())
	}

	if format == "jpeg" {
		reader := GetDecodeMethod(JPEG)
		fmt.Print(reader.getDecodeImage(image).toString())
	}
}

File: patterns/01_factory-method/ex2_ImageDecoder.java
// Step 1: Create a abstract class
// ===> Đây tập trung những method common, mà class khác đều phải có
interface ImageReader {
    DecodedImage getDecodeImage();
}

// Step 2: Create an algorithm to decode image
// ===> Hiểu rằng, việc một class decode, ta phải hiểu được
// có bao nhiêu parameters, các methods common nào cần thiết.
// Có trong class này, để ta provide cho những Decoder còn lại
// handle và implement riêng cho nó.
class DecodedImage {
    private String image;

    // Constructor
    public DecodedImage(String image) {
        this.image = image;
    }

    @Override
    public String toString() {
        return image + ": is decoded";
    }
}

// Step 3: Create a class to implement all interfaces
class GifReader implements ImageReader {
    private DecodedImage decodedImage;

    // Constructor
    public GifReader(String image) {
        this.decodedImage = new DecodedImage(image);
    }

    // Overview method decode for this class
    @Override
    public DecodedImage getDecodeImage() {
        return decodedImage;
    }
}

// Step 4: With new decoder,
//         we create another class to implement all interfaces
class JpegReader implements ImageReader {
    private DecodedImage decodedImage;

    public JpegReader(String image) {
        decodedImage = new DecodedImage(image);
    }

    @Override
    public DecodedImage getDecodeImage() {
        return decodedImage;
    }
}

public class FactoryMethodDemo {
    public static void main(String[] args) {
        // Initialize value
        DecodedImage decodedImage;
        ImageReader reader = null;
        String image = args[0];

        // Get format image from user
        String format = image.substring(image.indexOf('.') + 1, (image.length()));
        
        // Select extension image to start decoding image
        if (format.equals("gif")) {
            reader = new GifReader(image);
        }
        if (format.equals("jpeg")) {
            reader = new JpegReader(image);
        }
        assert reader != null;

        // Handle image
        decodedImage = reader.getDecodeImage();
        System.out.println(decodedImage);
    }
}